src/guest_memory.rs

```diff
@@ -646,10 +646,15 @@ pub trait GuestMemory {
                 Ok(0) => return Ok(total),
                 // made some progress
                 Ok(len) => {
-                    total += len;
-                    if total == count {
-                        break;
-                    }
+                    match total.checked_add(len) {
+                        Some(x) if x <= count => {
+                            total = x;
+                            if total == count {
+                                break;
+                            }
+                        },
+                        _ => return Err(Error::InvalidBackendAddress),
```

Comment:
I find this slightly awkward since the Backend Address is not actually invalid, but rather the access size. I think the major release problem is giving the wrong incentives in this case. I would prefer us having it as it was initially with a new error type, but I am also fine with disagreeing and committing on this one.

Reply:
While I suggested the error code, I got no opinion on this. Happy with either way.

Reply:
I've reverted it back to using a new specific error code.